quadtree/quadtree.go

```diff
@@ -0,0 +1,385 @@
+	freeNodes []node
+	freeIndex int
+}
+
+// node represents a node of the quad tree. Each node stores a Value
+// and has links to its 4 children
+type node struct {
+	children [4]*node
+	internal bool
+	pointer  geo.Pointer
+}
+
+// New creates a new quadtree for the given bound. Added points
+// must be within this bound.
+func New(bound *geo.Bound) *Quadtree {
+	return &Quadtree{
+		bound: bound,
+	}
+}
+
+// NewFromPointSet creates a quadtree from a pointset.
+// Copies the points into the quad tree. Modifying the points later
+// will invalidate the quad tree and lead to unexpected result.
+func NewFromPointSet(set *geo.PointSet) *Quadtree {
+	q := &Quadtree{
+		bound:     set.Bound(),
+		freeNodes: make([]node, set.Length(), set.Length()),
+	}
+
+	ps := []geo.Point(*set)
+	for i := range ps {
+		q.Insert(&ps[i])
+	}
+
+	return q
+}
+
+// NewFromPointers creates a quadtree from a set of pointers.
+func NewFromPointers(points []geo.Pointer) *Quadtree {
+	if len(points) == 0 {
+		// This is kind of meaningless but is what will happen
+		// if using an empty pointset above.
+		return &Quadtree{
+			bound: geo.NewBound(0, 0, 0, 0),
+		}
+	}
+
+	b := geo.NewBoundFromPoints(points[0].Point(), points[0].Point())
+	for _, p := range points {
+		b.Extend(p.Point())
+	}
+
+	q := &Quadtree{
+		bound:     b,
+		freeNodes: make([]node, len(points), len(points)),
+	}
+
+	for _, p := range points {
+		q.Insert(p)
+	}
+
+	return q
+}
+
+// Bound returns the bounds used for the quad tree.
+func (q *Quadtree) Bound() *geo.Bound {
+	return q.bound
+}
+
+// Insert puts an object into the quad tree, must be within the quadtree bounds.
+// If the pointer returns nil, the point will be ignored.
+// This function is not thread-safe, ie. multiple goroutines cannot insert into
+// a single quadtree.
+func (q *Quadtree) Insert(p geo.Pointer) error {
+	if p == nil {
+		return nil
+	}
+
+	point := p.Point()
+	if point == nil {
+		return nil
+	}
+
+	if !q.bound.Contains(point) {
+		return ErrPointOutsideOfBounds
+	}
+
+	if q.root == nil {
+		q.root = &node{}
+	}
+
+	q.insert(q.root, p,
+		q.bound.Left(), q.bound.Right(),
+		q.bound.Bottom(), q.bound.Top(),
+	)
+
+	return nil
+}
+
+// nextNode returns the next node from a preallocated list.
+// This resulted in about 15% improvement in quadtree creation.
+func (q *Quadtree) nextNode() *node {
+	if l := len(q.freeNodes); q.freeIndex >= l {
+		// Exponentially decrease the preallocation size.
+		// On a handful of tests, number of nodes was about 1.5 times pointers.
+		l /= 2
+
+		// min size of the preallocation. I think this could be bigger as it's
+		// not that much memory overhead. Optimizing this more would need
+		// to be use case specific.
+		if l < 25 {
+			l = 25
+		}
+
+		q.freeNodes = make([]node, l, l)
+		q.freeIndex = 0
+	}
+
+	n := &q.freeNodes[q.freeIndex]
+	q.freeIndex++
+	return n
+}
+
+func (q *Quadtree) insert(n *node, p geo.Pointer, left, right, bottom, top float64) {
+	point := p.Point()
```

Comment:
A short description of each case could provide some useful context for coming back to this code in the future.